library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -1,65 +1,216 @@
+
+-- | Compile the input program to an LLVM module.
+-- TODO: maybe do something smarter with the module name?
 compileProgram ::
   Monad m =>
+  -- | Term to compile.
   ErasedAnn.AnnTerm PrimTy RawPrimVal ->
   FeedbackT [] P.String m Text
-compileProgram t@(ErasedAnn.Ann usage ty _) = do
+compileProgram t = do
   let llvmmod :: LLVM.Module
-      llvmmod = LLVM.buildModule "juvix-module" $ do
-        LLVM.function "main" [] (typeToLLVM ty) $ \[] -> do
-          out <- compileTerm t
-          LLVM.ret out
+      llvmmod = LLVM.buildModule "juvix-module" $ mkMain t
   return $ toStrict $ LLVM.ppllvm llvmmod
 
+-- | Write the main function of the module. Here two distinct cases can be
+-- observed:
+--
+-- * The term @t@ is a lambda abstraction; this is the case when the main
+-- function takes one or more arguments. Within the scope of the input
+-- program, this lambda is never applied, as this only happens when the program
+-- is executed.
+-- A global function is written for the lambda and called in the main function.
+-- The arguments for the call are passed on from main.
+--
+-- * The term @t@ is any other term than a lambda abstraction, we can write the
+-- body of the main function by compiling @t@. The main function itself does
+-- not have any parameters in this case.
+mkMain ::
+  LLVM.MonadModuleBuilder m =>
+  -- | Term to compile.
+  ErasedAnn.AnnTerm PrimTy RawPrimVal ->
+  m LLVM.Operand
+mkMain t@(ErasedAnn.Ann usage ty t') = do
+  let types = map typeToLLVM (functionTy ty)
+      returnType = P.last types
+      paramTypes = init types
+      paramNames = repeat "arg"
+      params = zip paramTypes (map mkParameterName paramNames)
+  LLVM.function "main" params returnType $ \args -> do
+    out <- case t' of
+      ErasedAnn.LamM
+        { ErasedAnn.capture,
+          ErasedAnn.arguments,
+          ErasedAnn.body
+        } -> do
+          let env = Map.fromList $ zip paramNames args
+              callArgs = zip args (repeat []) -- No arg attributes.
+          funname <- mkLam env ty body arguments capture
+          LLVM.call (globalRef (typeToLLVM ty) funname) callArgs
+      _ -> compileTerm mempty t
+    LLVM.ret out
+
+-- | Compile a term to its equivalent LLVM code.
+-- TODO: implement other constructors of ErasedAnn.Term
 compileTerm ::
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of Juvix variables to LLVM function arguments.
+  Env ->
+  -- | The term to compile.
   ErasedAnn.AnnTerm PrimTy RawPrimVal ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-compileTerm (ErasedAnn.Ann usage ty t) = case t of
-  ErasedAnn.Prim t' -> mkPrim t
-  ErasedAnn.AppM f xs -> mkApp (ErasedAnn.term f) ty xs
+  m LLVM.Operand
+compileTerm env (ErasedAnn.Ann usage ty t) = case t of
+  ErasedAnn.Var symbol -> case Map.lookup symbol env of
+    Nothing ->
+      P.error $ "Variable not found: " <> show symbol <> " in " <> show env -- TODO improve error message.
```

Comment:
we should use something other than Prelude.error I think?

Reply:
or is this a compiler fault?

Reply:
I am not a fan of using error, but it is a quick solution for now.

This case should actually never occur, as a program referring to an unknown variable shouldn't typecheck. For now this problem occurs when using a global variable that we do not have access to in the backend.